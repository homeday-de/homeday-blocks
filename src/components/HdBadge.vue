<template>
  <!-- collapsible badge component could have been structured in a semantic way using
  html5 <details> & <summary> tags but unfortunately IE doesn't support them -->
  <div
    :class="computedClasses"
  >
    <span
      class="badge__label"
      @click.capture="collapse"
    >
      <template v-if="showIconBefore">
        <img
          v-if="icon"
          :src="icon"
          class="badge__icon"
        >
        <svg
          v-else
          class="badge__icon"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="7"
            cy="7"
            r="7"
            fill="currentColor"
          />
        </svg>
      </template>
      {{ label }}
      <HdIcon
        :src="chevronIcon"
        transform="rotate(90)"
        class="badge__chevron"
      />
    </span>
    <p
      v-if="hasDetails"
      class="badge__details"
    >
      <slot />
    </p>
  </div>
</template>

<script>
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import { chevronIcon } from 'homeday-blocks/src/assets/small-icons';

export const TYPES = {
  DEFAULT: '',
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  DISABLED: 'disabled',
};

export default {
  name: 'HdBadge',
  components: {
    HdIcon,
  },
  props: {
    modifier: {
      type: String,
      default: TYPES.DEFAULT,
      validator(value) {
        const allTypes = Object.values(TYPES);

        return allTypes.includes(value);
      },
    },
    label: {
      type: String,
      default: '',
    },
    showIconBefore: {
      type: Boolean,
      default: true,
    },
    icon: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      collapsed: false,
      chevronIcon,
    };
  },
  computed: {
    computedClasses() {
      const baseClass = 'badge';
      const classes = [baseClass];

      if (this.modifier) {
        classes.push(`${baseClass}--${this.modifier}`);
      }
      if (this.hasDetails) {
        classes.push(`${baseClass}--with-details`);
      }
      if (this.collapsed) {
        classes.push(`${baseClass}--collapsed`);
      }

      return classes;
    },
    hasDetails() {
      return !!this.$slots.default;
    },
  },
  methods: {
    collapse() {
      if (this.hasDetails) this.collapsed = !this.collapsed;
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'homeday-blocks/src/styles/mixins.scss';

.badge{
    display: flex;
    justify-content: flex-start;

    &__label{
      display: flex;
      align-items: center;
      font-weight: bold;
      font-size: 18px;
      line-height: 28px;
      border-radius: 4px;
      padding: $stack-xs $inline-s;
      cursor: default;
    }

    &__chevron {
      display: none;
      width: 24px;
      height: 24px;
      margin-left: $inline-m;
      margin-right: $inline-s;

      &::v-deep path {
        fill: currentColor;
      }
    }

    ::v-deep  &__icon{
      width: 14px;
      height: 14px;
      margin-right: $inline-s;
    }

    &__details{
      visibility: hidden;
      width: 170px;
      white-space: nowrap;
      overflow: hidden;

      padding: $inset-s;
      padding-top: 0;
      border-bottom-left-radius: $inline-xs;
      border-bottom-right-radius: $inline-xs;
      border-top-left-radius: 0;
      border-top-right-radius: 0;
      font-weight: normal;
      font-size: 18px;
      line-height: 28px;
    }

    &--primary{
        color: $success-color;

        .badge__label, .badge__details{
          background: rgba($success-color, 0.15);
        }
    }

    &--secondary{
        color: getShade($secondary-color, 110);

        .badge__label, .badge__details{
          background: getShade($secondary-color, 70);
        }
    }

    &--tertiary{
        color: $tertiary-color;

        .badge__label, .badge__details{
          background: getShade($tertiary-color, 60);
        }
    }

    &--disabled{
        color: $quaternary-color;

        .badge__label, .badge__details{
          background: getShade($quaternary-color, 40);
        }
    }

    &--with-details & {
      &__label{
        cursor: pointer;
      }

      &__chevron{
        display: block;
      }
    }

    &--collapsed {
      display: block;
    }

    &--collapsed > &__label {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }

    &--collapsed &__chevron {
      transform: rotateZ(270deg);
    }

    &--collapsed > &__details {
      visibility: visible;
      width: 100%;
      white-space: normal;
    }
}
</style>
