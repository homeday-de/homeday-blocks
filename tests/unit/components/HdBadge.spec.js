import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdBadge, { TYPES } from '@/components/HdBadge.vue';

const LABEL_SELECTOR = '.badge__label';
const DETAILS_SELECTOR = '.badge__details';
const ICON_SELECTOR = '.badge__icon';

const BADGE_WITH_DETAILS_CLASS = 'badge--with-details';
const BADGE_COLLAPSED_CLASS = 'badge--collapsed';

const defaultSlot = 'test <strong>badge</strong> details!';
const defaultLabel = 'test label';

const wrapperBuilder = wrapperFactoryBuilder(HdBadge, {
  props: {
    label: defaultLabel,
  },
  slots: {
    default: defaultSlot,
  },
  shallow: true,
});

describe('HdBadge', () => {
  it('should render', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.classes()).toContain('badge');
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('modifier variations', () => {
    it('should render with `badge--primary` class if modifier is set to primary', () => {
      const modifier = TYPES.PRIMARY;
      const className = `badge--${modifier}`;
      const wrapper = wrapperBuilder({
        props: {
          modifier,
        },
      });

      expect(wrapper.classes()).toContain(className);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render with `badge--secondary` class if modifier is set to secondary', () => {
      const modifier = TYPES.SECONDARY;
      const className = `badge--${modifier}`;
      const wrapper = wrapperBuilder({
        props: {
          modifier,
        },
      });

      expect(wrapper.classes()).toContain(className);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render with `badge--tertiary` class if modifier is set to tertiary', () => {
      const modifier = TYPES.TERTIARY;
      const className = `badge--${modifier}`;
      const wrapper = wrapperBuilder({
        props: {
          modifier,
        },
      });

      expect(wrapper.classes()).toContain(className);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render with `badge--disabled` class if modifier is set to disabled', () => {
      const modifier = TYPES.DISABLED;
      const className = `badge--${modifier}`;
      const wrapper = wrapperBuilder({
        props: {
          modifier,
        },
      });

      expect(wrapper.classes()).toContain(className);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('should render label correctly', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.find(LABEL_SELECTOR).element.textContent.trim()).toBe(defaultLabel);
  });

  it('should render details correctly', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.classes()).toContain(BADGE_WITH_DETAILS_CLASS);
    expect(wrapper.find(DETAILS_SELECTOR).element.innerHTML).toBe(defaultSlot);
  });

  it('should collapse details if the label was clicked', async () => {
    const wrapper = wrapperBuilder();

    wrapper.find(LABEL_SELECTOR).element.click();

    await wrapper.vm.$nextTick();

    expect(wrapper.classes()).toContain(BADGE_COLLAPSED_CLASS);
  });

  it('should not render details if a slot wasn\'t provided', () => {
    const wrapper = wrapperBuilder({
      slots: {
        default: '',
      },
    });

    expect(wrapper.classes()).not.toContain(BADGE_WITH_DETAILS_CLASS);
    expect(wrapper.find(DETAILS_SELECTOR).exists()).toBeFalsy();
  });

  it('should render a default icon', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.find(ICON_SELECTOR).exists()).toBeTruthy();
  });

  it('should render a custom icon', () => {
    const ICON_PATH = 'fake/icon.svg';
    const wrapper = wrapperBuilder({
      props: {
        icon: ICON_PATH,
      },
    });

    expect(wrapper.find(ICON_SELECTOR).attributes().src).toBe(ICON_PATH);
  });

  it('should not render an icon if the prop `show-icon-before` is set false', () => {
    const wrapper = wrapperBuilder({
      props: {
        showIconBefore: false,
      },
    });

    expect(wrapper.find(ICON_SELECTOR).exists()).toBeFalsy();
  });
});
