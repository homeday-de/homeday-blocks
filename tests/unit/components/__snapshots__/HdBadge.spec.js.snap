// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdBadge modifier variations should render with \`badge--disabled\` class if modifier is set to disabled 1`] = `
<div class="badge badge--disabled badge--with-details"><span class="badge__label"><svg xmlns="http://www.w3.org/2000/svg" class="badge__icon"><circle cx="7" cy="7" r="7" fill="currentColor"></circle></svg>
    test label
    <hdicon-stub src="" transform="rotate(90)" class="badge__chevron"></hdicon-stub></span>
  <p class="badge__details">test <strong>badge</strong> details!</p>
</div>
`;

exports[`HdBadge modifier variations should render with \`badge--primary\` class if modifier is set to primary 1`] = `
<div class="badge badge--primary badge--with-details"><span class="badge__label"><svg xmlns="http://www.w3.org/2000/svg" class="badge__icon"><circle cx="7" cy="7" r="7" fill="currentColor"></circle></svg>
    test label
    <hdicon-stub src="" transform="rotate(90)" class="badge__chevron"></hdicon-stub></span>
  <p class="badge__details">test <strong>badge</strong> details!</p>
</div>
`;

exports[`HdBadge modifier variations should render with \`badge--secondary\` class if modifier is set to secondary 1`] = `
<div class="badge badge--secondary badge--with-details"><span class="badge__label"><svg xmlns="http://www.w3.org/2000/svg" class="badge__icon"><circle cx="7" cy="7" r="7" fill="currentColor"></circle></svg>
    test label
    <hdicon-stub src="" transform="rotate(90)" class="badge__chevron"></hdicon-stub></span>
  <p class="badge__details">test <strong>badge</strong> details!</p>
</div>
`;

exports[`HdBadge modifier variations should render with \`badge--tertiary\` class if modifier is set to tertiary 1`] = `
<div class="badge badge--tertiary badge--with-details"><span class="badge__label"><svg xmlns="http://www.w3.org/2000/svg" class="badge__icon"><circle cx="7" cy="7" r="7" fill="currentColor"></circle></svg>
    test label
    <hdicon-stub src="" transform="rotate(90)" class="badge__chevron"></hdicon-stub></span>
  <p class="badge__details">test <strong>badge</strong> details!</p>
</div>
`;

exports[`HdBadge should render 1`] = `
<div class="badge badge--with-details"><span class="badge__label"><svg xmlns="http://www.w3.org/2000/svg" class="badge__icon"><circle cx="7" cy="7" r="7" fill="currentColor"></circle></svg>
    test label
    <hdicon-stub src="" transform="rotate(90)" class="badge__chevron"></hdicon-stub></span>
  <p class="badge__details">test <strong>badge</strong> details!</p>
</div>
`;
